torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -339,13 +458,28 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::broadcast(
     entry->src[opts.rootTensor].copy_(tensors[opts.rootTensor]);
   }
 
-  // Define how to run the algorithm and copy back results
-  entry->run = [=]() mutable {
-    entry->algorithm->run();
-    for (int i = 0; i < tensors.size(); i++) {
-      tensors[i].copy_(entry->src[i]);
-    }
-  };
+  // In case of CUDA, ensure that operations that are queued after
+  // this collective wait for the collective to complete.
+  if (key.type->is_cuda()) {
+    synchronizeStreams(thcState_, entry);
```

Comment:
Shouldn't this be part of the `run` body? Why do we only wait for public streams only once?

Reply:
We need to capture the state of the caller's streams at the call site. Everything in the run body is run by a different thread, where we have a different thread local streams.